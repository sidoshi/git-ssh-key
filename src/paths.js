import path from 'path'

export default {
  tmp: path.resolve('tmp'),
  keysDir: path.resolve('tmp/keys'),
  configDir: path.resolve('tmp/config'),
}
