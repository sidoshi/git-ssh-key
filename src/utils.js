import fs from 'fs-extra'
import paths from './paths'

export const delimeter = '### git-ssh-key-config ###'

export const base64Decode = (encoded: string) =>
  Buffer.from(encoded, 'base64').toString('utf8')

export const cleanupConfig = (config: string) => {
  const start = config.indexOf(delimeter)
  const end = config.indexOf(delimeter, start + 1) + delimeter.length

  if (start < 0) return config.trim()

  const newConfig = (config.substring(0, start) + config.substring(end)).trim()
  return cleanupConfig(newConfig)
}

export const cleanup = () => {
  // cleanup keys
  fs.removeSync(paths.keysDir)

  if (!fs.existsSync(paths.configFile)) return

  // cleanup config
  const config = fs.readFileSync(paths.configFile).toString()
  const cleanedConfig = cleanupConfig(config)
  cleanedConfig.length === 0
    ? fs.removeSync(paths.configFile)
    : fs.writeFileSync(paths.configFile, cleanedConfig)
}
