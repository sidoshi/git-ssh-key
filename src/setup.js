import path from 'path'
import fs from 'fs-extra'
import chalk from 'chalk'

import { base64Decode, cleanup } from './utils'
import paths from './paths'

import type { PlatformKeys, KeysEnv } from './types'

const mapPlatformToUrl = {
  github: 'github.com',
  gitlab: 'gitlab.com',
  bitbucket: 'bitbucket.org',
}

export const delimeter = '### git-ssh-key-config ###'

export const filterKeys = (keys: PlatformKeys) =>
  Object.keys(keys).reduce((validKeys, platform) => {
    if (!keys[platform]) return validKeys

    validKeys[platform] = keys[platform]
    return validKeys
  }, {})

export const getKeyPath = (platform: string) =>
  path.resolve(paths.keysDir, `${platform}_key`)

export const buildConfig = (platform: string) =>
  `${delimeter}
Host ${mapPlatformToUrl[platform]}
  IdentitiesOnly yes
  UserKnownHostsFile=/dev/null
  StrictHostKeyChecking no
  IdentityFile ${getKeyPath(platform)}
${delimeter}
`

export const createKeyFile = (platform: string, key: string) => {
  fs.ensureDirSync(paths.keysDir)
  const keyPath = getKeyPath(platform)
  fs.writeFileSync(keyPath, key)
  fs.chmodSync(keyPath, '600')
}

export const appendConfigFile = (platform: string) =>
  fs.appendFileSync(paths.configFile, buildConfig(platform))

export const setupSshConfig = (keys: PlatformKeys) => {
  const validKeys = filterKeys(keys)

  Object.keys(validKeys).forEach(platform => {
    const key = validKeys[platform]
    createKeyFile(platform, key)
    appendConfigFile(platform)
  })
}

export default (env: KeysEnv) => {
  cleanup()

  const {
    GIT_SSH_KEY_GITHUB: githubKey,
    GIT_SSH_KEY_GITLAB: gitlabKey,
    GIT_SSH_KEY_BITBUCKET: bitbucketKey,
  } = env

  // If a private key of a specific platform is provided, setup env using
  // configs to allow all platforms.
  if (githubKey || gitlabKey || bitbucketKey) {
    const keys = {
      github: base64Decode(githubKey),
      gitlab: base64Decode(gitlabKey),
      bitbucket: base64Decode(bitbucketKey),
    }

    setupSshConfig(keys)
    return
  }

  console.log(
    chalk.red(`
You must set private keys in appropriate environment variables, either manually 
or in a '.env' file encoded using base64 algorithm.

git-ssh-key reads from the folowing environment variables :-

GIT_SSH_KEY_GITHUB
GIT_SSH_KEY_BITBUCKET
GIT_SSH_KEY_GITLAB
  `)
  )

  process.exit(1)
}

/* eslint no-param-reassign: off */
