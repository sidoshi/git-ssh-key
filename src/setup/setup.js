import path from 'path'
import fs from 'fs-extra'

import { delimeter } from '../utils'
import paths from '../paths'

type PrivateKey = string
type Host = string

export const buildConfig = (host: Host, identityFile: string) => `
${delimeter}
Host ${host}
  IdentitiesOnly yes
  UserKnownHostsFile=/dev/null
  StrictHostKeyChecking no
  IdentityFile ${identityFile}
${delimeter}
`

export const getPrivateKeyFilePath = (host: Host) =>
  path.resolve(paths.keysDir, `${host.replace(/\./g, '_')}_key`)

export const appendConfigFile = (host: Host) =>
  fs.appendFileSync(
    paths.configFile,
    buildConfig(host, getPrivateKeyFilePath(host))
  )

export const createKeyFile = (host: string, key: string) => {
  fs.ensureDirSync(paths.keysDir)
  const privateKeyFilePath = getPrivateKeyFilePath(host)

  fs.writeFileSync(privateKeyFilePath, key)
  fs.chmodSync(privateKeyFilePath, '600')
}

export default (pairs: [Host, PrivateKey][]) =>
  pairs.forEach(([host, key]) => {
    createKeyFile(host, key)
    appendConfigFile(host)
  })
