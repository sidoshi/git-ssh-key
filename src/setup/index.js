import setup from './setup'
import { cleanup, base64Decode } from '../utils'

const keyNotFoundMessage = () =>
  `
Private keys not found.

You must set private keys in appropriate environment variables, either manually
or in a '.env' file, encoded using base64 algorithm.

git-ssh-key reads private keys from environment variables matching pattern:

GIT_SSH_KEY_XXXX

For eg:-

GIT_SSH_KEY_GITHUB
GIT_SSH_KEY_BITBUCKET
GIT_SSH_KEY_GITLAB
 `

const hostNotFoundMessage = (keyName, hostName) => `
Host url not found for corresponding private key: ${keyName}
Missing environment variable: ${hostName}
      
You must set host urls in appropriate environment variables, either manually
or in a '.env' file.

git-ssh-key reads host urls from environment variables matching pattern:

GIT_SSH_HOST_XXXX

You dont need to provide host urls for a few popular git services like:
Github, Gitlab and Bitbucket. Though If you do provide host urls for those 
services, it will override the default values.

For eg:-

GIT_SSH_HOST_COMPANY_GITLAB
GIT_SSH_HOST_GITHUB
GIT_SSH_HOST_BITBUCKET
`

export const getKeyNames = (envs: Object) =>
  Object.keys(envs).filter(key => key.match(/GIT_SSH_KEY_.+/))

export const getHosts = (envs: Object) => ({
  GIT_SSH_HOST_GITHUB: 'github.com',
  GIT_SSH_HOST_BITBUCKET: 'bitbucket.org',
  GIT_SSH_HOST_GITLAB: 'gitlab.com',
  ...Object.keys(envs).reduce(
    (acc, curr) =>
      curr.match(/GIT_SSH_HOST_.+/) ? { ...acc, [curr]: envs[curr] } : acc,
    {}
  ),
})

export default (envs: Object) => {
  cleanup()

  const keyNames = getKeyNames(envs)
  if (keyNames.length === 0) {
    console.log(keyNotFoundMessage())
    process.exit()
  }

  const hosts = getHosts(envs)

  const pairs = keyNames.map(keyName => {
    const privateKey = base64Decode(envs[keyName])

    const hostName = keyName.replace('GIT_SSH_KEY', 'GIT_SSH_HOST')
    const host = hosts[hostName]
    if (!host) {
      console.log(hostNotFoundMessage(keyName, hostName))
      process.exit(1)
    }

    return [host, privateKey]
  })

  setup(pairs)
}
