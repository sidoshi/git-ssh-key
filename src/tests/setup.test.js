import path from 'path'
import fs from 'fs-extra'

import setup, {
  filterKeys,
  setupSshConfig,
  buildConfig,
  getKeyPath,
  createKeyFile,
  appendConfigFile,
} from '../setup'
import { cleanup } from '../utils'
import paths from '../paths'

const _exit = process.exit
const _log = console.log

beforeEach(() => {
  process.exit = jest.fn()
  console.log = jest.fn()
  cleanup()
})

afterAll(() => {
  process.exit = _exit
  console.log = _log
  cleanup()
})

test('Filters keys properly', () => {
  expect(
    filterKeys({ github: null, gitlab: '', bitbucket: undefined })
  ).toEqual({})
  expect(
    filterKeys({
      github: 'github',
      gitlab: 'something',
      bitbucket: undefined,
    })
  ).toEqual({ github: 'github', gitlab: 'something' })
})

test('getKeyPath', () => {
  expect(getKeyPath('github')).toMatch('.ssh/git_ssh_keys/github_key')
})

test('BuildConfig', () => {
  expect(buildConfig('github')).toMatch('github.com')
  expect(buildConfig('github')).toMatch('github_key')
  expect(buildConfig('github')).toMatch('### git-ssh-key-config ###')
  expect(buildConfig('gitlab')).toMatch('gitlab.com')
  expect(buildConfig('gitlab')).toMatch('gitlab_key')
  expect(buildConfig('bitbucket')).toMatch('bitbucket.org')
  expect(buildConfig('bitbucket')).toMatch('bitbucket_key')
})

test('creates key file properly', () => {
  const keyPath = path.resolve(paths.keysDir, 'github_key')
  expect(fs.existsSync(keyPath)).toBe(false)
  createKeyFile('github', 'githubKey')
  expect(fs.existsSync(keyPath)).toBe(true)
  expect(fs.readFileSync(keyPath).toString()).toBe('githubKey')
  expect(fs.statSync(keyPath).mode).toBe(33152)
})

test('Appends config file properly', () => {
  const configPath = paths.configFile
  appendConfigFile('github')
  expect(fs.existsSync(configPath)).toBe(true)
  expect(fs.readFileSync(configPath).toString()).toMatch('github.com')
  expect(fs.readFileSync(configPath).toString()).toMatch('github_key')
  appendConfigFile('gitlab')
  expect(fs.readFileSync(configPath).toString()).toMatch('gitlab.com')
  expect(fs.readFileSync(configPath).toString()).toMatch('gitlab_key')
})

test('Setups config file properly when platform keys are provided', () => {
  const env = {}
  setupSshConfig(
    {
      github: 'githubKey',
      gitlab: 'gitlabKey',
      bitbucket: null,
    },
    env
  )
  expect(fs.existsSync(path.resolve(paths.keysDir, 'github_key'))).toBe(true)
  expect(fs.existsSync(path.resolve(paths.keysDir, 'gitlab_key'))).toBe(true)
  expect(fs.existsSync(path.resolve(paths.keysDir, 'bitbucket_key'))).toBe(
    false
  )
  expect(fs.existsSync(paths.configFile)).toBe(true)
  const config = fs.readFileSync(path.resolve(paths.configFile)).toString()
  expect(config).toMatch('gitlab_key')
  expect(config).toMatch('github_key')
  expect(config).toMatch('gitlab.com')
  expect(config).toMatch('github.com')
  expect(config).not.toMatch('bitbucket_key')
  expect(config).not.toMatch('bitbucket.org')
})

test('Shows error message if no keys are found', () => {
  setup({})
  expect(console.log.mock.calls[0][0]).toMatchSnapshot()
  expect(process.exit).toBeCalledWith(1)
})

test('Works if platform key is provided', () => {
  setup({ GIT_SSH_KEY_GITHUB: 'github' })
  expect(process.exit).not.toBeCalledWith(1)
})
