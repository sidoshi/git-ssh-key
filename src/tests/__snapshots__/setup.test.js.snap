// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shows error message if no keys are found 1`] = `
"[31m[39m
[31mYou must set private keys in appropriate environment variables, either manually [39m
[31mor in a '.env' file encoded using base64 algorithm.[39m
[31m[39m
[31mgit-ssh-key reads from the folowing environment variables :-[39m
[31m[39m
[31mGIT_SSH_KEY[39m
[31mGIT_SSH_KEY_GITHUB[39m
[31mGIT_SSH_KEY_BITBUCKET[39m
[31mGIT_SSH_KEY_GITLAB[39m
[31m  [39m"
`;
